molecule.go

```diff
@@ -149,30 +199,84 @@ func (m Molecule) Mount(dest string) error {
 		return errors.WithStack(err)
 	}
 
-	mntOpts, err := m.overlayArgs(dest)
+	overlayLowerDirs, err := m.overlayLowerDirs()
 	if err != nil {
 		return err
 	}
 
-	// The kernel doesn't allow mount options longer than 4096 chars, so
-	// let's give a nicer error than -EINVAL here.
-	if len(mntOpts) > 4096 {
-		return errors.Errorf("too many lower dirs; must have fewer than 4096 chars")
-	}
+	complete := false
 
-	err = m.mountUnderlyingAtoms()
+	defer func() {
+		if !complete {
+			log.Errorf("Failure detected: cleaning up %q", metadir)
+			os.RemoveAll(metadir)
+		}
+	}()
+
+	err, cleanupUnderlyingAtoms := m.mountUnderlyingAtoms()
 	if err != nil {
 		return err
 	}
 
-	err = m.config.WriteToFile(filepath.Join(m.config.MetadataPath, "config.json"))
+	defer func() {
+		if !complete {
+			cleanupUnderlyingAtoms()
+		}
+	}()
+
+	err = m.config.WriteToFile(filepath.Join(metadir, "config.json"))
 	if err != nil {
 		return err
 	}
 
-	// now, do the actual overlay mount
-	err = unix.Mount("overlay", dest, "overlay", 0, mntOpts)
-	return errors.Wrapf(err, "couldn't do overlay mount to %s, opts: %s", dest, mntOpts)
+	overlayArgs := ""
+	if m.config.AddWriteableOverlay {
+		rodest := filepath.Join(metadir, "ro")
+		if err = EnsureDir(rodest); err != nil {
+			return err
+		}
+
+		workdir := filepath.Join(metadir, "work")
+		if err := EnsureDir(workdir); err != nil {
+			return errors.Wrapf(err, "failed to ensure workdir")
+		}
+
+		upperdir := m.config.WriteableOverlayPath
+		if upperdir == "" {
+			upperdir = filepath.Join(metadir, "persist")
```

Comment:
Hm, I think this is an existing "bug" in the code you're moving, but note that workdir and upperdir must be on the same filesystem. Therefore, it might be best to have workdir be created adjecent to the persist dir, right?

Reply:
ah, thanks. I didn't know that. yeah this was in the original code but we might as well fix it now